File: linkedList/implementation.cpp
#include <iostream>
using namespace std;

class Node
{
public:
  int data;
  Node *next;
};

// Used double pointer here for the 'head_ref' as the 'next' param in class 'Node' is a pointer type.
// So if we want to get the address of the head in the last Node then we will use a pointer to the 'next' pointer
// We can then dereference the double pointer to eventually get the address of the last Node object

// Adds a new item at the start of the Linked List
void push(Node **head_ref, int new_value)
{
  Node *new_node = new Node();
  new_node->data = new_value;
  new_node->next = *head_ref;
  *head_ref = new_node;
}

// Inserts a new item at a specific location in the Linked List
void insertAt(Node *prev_node, int new_value)
{
  if (prev_node == NULL)
  {
    cout << "The previous node seems to be NULL" << endl;
    return;
  }
  Node *new_node = new Node();
  new_node->data = new_value;
  new_node->next = prev_node->next;
  prev_node->next = new_node;
}

// Inserts the new item at the end of the Linked List
void append(Node **head_ref, int new_value)
{
  Node *new_node = new Node();
  new_node->data = new_value;
  new_node->next = NULL;

  Node *last = *head_ref;

  if (*head_ref == NULL)
  {
    *head_ref = new_node;
    return;
  }

  while (last->next != NULL)
  {
    last = last->next;
  }

  last->next = new_node;
}

// Prints all the Linked List items
void printList(Node *node)
{
  while (node != NULL)
  {
    cout << node->data << " ";
    node = node->next;
  }
  cout << "\n";
}

// Deletes an element from the Linked List with a specific key
void deleteItem(Node **head_ref, int key)
{
  Node *curr_node = *head_ref, *prev;

  if (*head_ref == NULL)
  {
    cout << "This is an empty linked list" << endl;
    return;
  }

  // Case 1: Deleting the first element of the Linked List
  if (curr_node != NULL && curr_node->data == key)
  {
    *head_ref = curr_node->next;
    free(curr_node);
    return;
  }

  // Case 2: Deleting the node anywhere other than the one at the head_ref
  while (curr_node != NULL && curr_node->data != key)
  {
    prev = curr_node;
    curr_node = curr_node->next;
  }

  // Case 3: If the element is not present in the Linked List
  if (curr_node == NULL)
  {
    cout << "There is no element in the Linked List with the key: " << key << endl;
    return;
  }

  // Set the 'next' of the prev to that of curr_node;
  prev->next = curr_node->next;
  // Freeing up the Node referenced by the 'curr_node' pointer
  free(curr_node);
}

void deleteEntireList(Node **head_ref)
{
  Node *curr_node = *head_ref, *next_node;

  while (curr_node != NULL)
  {
    next_node = curr_node->next;
    free(curr_node);
    curr_node = next_node;
  }

  // Reset the 'head_ref' pointer to NULL
  *head_ref = NULL;
}

int getLength(Node *head_ref)
{
  int count = 0;
  Node *curr_node = head_ref;

  while (curr_node != NULL)
  {
    curr_node = curr_node->next;
    count++;
  }
  return count;
}

int testing()
{
  // Setting the initial head pointer to point to NULL as
  // there are no items present in the Linked List yet
  Node *head = NULL;

  // Insering new items into the Linked List
  append(&head, 11);
  append(&head, 12);
  append(&head, 13);
  append(&head, 14);
  push(&head, 12);
  printList(head);
  cout << "The length of the Linked List is: " << getLength(head) << endl;

  // Inserting new item at a specific location in the Linked List
  insertAt(head, 15);
  printList(head);
  cout << "The length of the Linked List is: " << getLength(head) << endl;

  // Deleting a single item from the Linked List
  // deleteItem(&head, 15);
  deleteItem(&head, 12);
  printList(head);
  cout << "The length of the Linked List is: " << getLength(head) << endl;

  // Delete the entire Linked List
  deleteEntireList(&head);
  printList(head);
  cout << "The length of the Linked List is: " << getLength(head) << endl;

  return 0;
}

File: linkedList/implementation.js
class Node {
  constructor(data){
    this.data = data;
    this.next = null;
  }

}


class LinkedList{
  root = null;

  constructor(rootNodeData){

    if(rootNodeData === undefined){
      return;
    }

    this.root = new Node(rootNodeData);
  }

  append(data){
    const newNode = new Node(data);

    if(this.root === null){
      this.root = newNode;
      return;
    }

    let curr = this.root;

    while(curr.next !== null){
      curr = curr.next;
    }

    curr.next = newNode;
  }

  push(data){
    const newNode = new Node(data);

    if(this.root === null){
      this.root = newNode;
      return;
    }

    newNode.next = this.root;
    this.root = newNode; 
  }

  reverse(){
    if(this.root === null || this.root.next === null){
      return;
    }

    let curr = this.root;
    let prevNode = null;
    let nextNode = null;
    
    while(curr.next !== null){
      nextNode = curr.next;
      curr.next = prevNode;
      prevNode = curr;
      curr = nextNode;
    }

    curr.next = prevNode;
    this.root = curr;
  }

  print(){
    if(this.root === null){
      return;
    }

    let curr = this.root;
    const result = [];

    while(curr.next !== null){
      result.push(curr.data);
      curr = curr.next;
    }
    
    result.push(curr.data);
    console.info(...result);
  }

  get length(){
    if(this.root === null){
      return 0;
    }

    if(this.root.next === null){
      return 1;
    }

    let curr = this.root;
    let result = 0;

    while(curr !== null){
      result++;
      curr = curr.next;
    }

    return result;
  }

  insertAt(position, data){

    if(position < 1 || position > this.length + 1){
      throw new Error('Position must be between 1 and list\'s length + 1');
    }

    if(position === 1){
      this.push(data);
      return;
    }

    if(position === this.length + 1){
      this.append(data);
      return;
    }


    let curr = this.root;
    let pos = 1;
    let prev = null;

    while(curr.next !== null){
      if(pos === position){
        const newNode = new Node(data);
        newNode.next = prev.next;
        prev.next = newNode;
        return;
      }

      pos++;
      prev = curr;
      curr = curr.next;
    }

    if(pos === position){
      const newNode = new Node(data);
      newNode.next = prev.next;
      prev.next = newNode;
    }
  }

}

(() => {
  const ll = new LinkedList(1);
  ll.append(2);
  ll.append(3);
  ll.append(4);
  ll.append(5);
  ll.append(6);
  ll.append(7);
  ll.append(8);
  ll.append(9);

  ll.print();

  // ll.push(0);
  ll.append(10);

  ll.print();
  
  // ll.reverse();

  ll.insertAt(5, 100);
  
  ll.print();
  console.log({length: ll.length})
})()